macos/library/FLEViewController.m

```diff
@@ -227,28 +220,45 @@ - (BOOL)launchEngineWithMainPath:(NSURL *)main
 }
 
 - (BOOL)addPlugin:(id<FLEPlugin>)plugin {
-  NSString *channel = plugin.channel;
-  if (_plugins[channel] != nil) {
-    NSLog(@"Warning: channel %@ already has an associated plugin", channel);
+  NSString *channelName = plugin.channel;
+  if (_messageHandlers[channelName] != nil) {
+    NSLog(@"Warning: channel %@ already has an associated plugin", channelName);
     return NO;
   }
-  _plugins[channel] = plugin;
+
   plugin.controller = self;
+  // See the note on the |codec| property of FLEPlugin for the reason this defaults to JSON.
+  id<FLEMethodCodec> codec = [FLEJSONMethodCodec sharedInstance];
+  if ([plugin respondsToSelector:@selector(codec)]) {
+    codec = plugin.codec;
+  }
+
+  // TODO: Move the channel into the plugin implementations, matching Flutter's iOS plugin
+  // structure. This is is a temporary shim to add FLEMethodChannel without coupling it to the
+  // breaking API change.
+  FLEMethodChannel *channel = [FLEMethodChannel methodChannelWithName:channelName
+                                                      binaryMessenger:self
+                                                                codec:codec];
+  // This will end up calling back to setMessageHandlerOnChannel:binaryMessageHandler: below, which
+  // retains the handler, so the strong reference to |plugin| here will keep the plugin alive.
+  [channel setMethodCallHandler:^(FLEMethodCall *call, FLEMethodResult result) {
+    [plugin handleMethodCall:call result:result];
+  }];
+
   return YES;
 }
 
 - (void)invokeMethod:(NSString *)method
            arguments:(id)arguments
-           onChannel:(NSString *)channel
+           onChannel:(NSString *)channelName
            withCodec:(id<FLEMethodCodec>)codec {
-  FLEMethodCall *methodCall = [[FLEMethodCall alloc] initWithMethodName:method arguments:arguments];
-  NSData *messageData = [codec encodeMethodCall:methodCall];
```

Comment:
I know it's not the point of this PR, but where is this object released?

Reply:
`messageData`? Nothing keeps a reference to it, so it'll be released when it goes out of scope. The engine doesn't require that the underlying bytes stay valid beyond the call to FlutterEngineSendPlatformMessage.